ide/gsf.testrunner.ui/src/org/netbeans/modules/gsf/testrunner/ui/api/TestResultDisplayHandler.java

```diff
@@ -0,0 +1,101 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.gsf.testrunner.ui.api;
+
+import org.netbeans.modules.gsf.testrunner.api.Report;
+import org.netbeans.modules.gsf.testrunner.api.TestSession;
+import org.netbeans.modules.gsf.testrunner.api.TestSuite;
+import org.netbeans.modules.gsf.testrunner.ui.ResultDisplayHandler;
+import org.openide.util.Lookup;
+
+/**
+ * Common base class for handlers displaying test results.
+ *
+ * @since 1.22
+ * @author Dusan Balek
+ */
+public abstract class TestResultDisplayHandler {
+
+    /**
+     * Get the {@link TestResultDisplayHandler} for the test session.
+     *
+     * @param session test session
+     * @return {@link TestResultDisplayHandler} instance
+     */
+    public static final TestResultDisplayHandler get(TestSession session) {
+        TestResultDisplayHandler.Provider provider = Lookup.getDefault().lookup(TestResultDisplayHandler.Provider.class);
+        if (provider != null) {
+            return provider.create(session);
+        }
+        return new ResultDisplayHandler(session);
+    }
+
+    public abstract void displayOutput(String text, boolean error);
+
+    /**
+     * Display information that a test suite is running.
+     *
+     * @param suiteName name of the running suite; or {@code null} in the case
+     *                  of anonymous suite
+     */
+    public abstract void displaySuiteRunning(String suiteName);
+
+    /**
+     * Display information that a test suite is running.
+     *
+     * @param suite the running suite
+     */
+    public abstract void displaySuiteRunning(TestSuite suite);
+
+    /**
+     * Display test results.
+     *
+     * @param report summary report to display
+     */
+    public abstract void displayReport(Report report);
+
+    /**
+     * Display message produced by running test.
+     *
+     * @param message message to display
+     */
+    public abstract void displayMessage(String message);
+
+    /**
+     * Display information that a test session has finished.
+     *
+     * @param message message to display
+     */
+    public abstract void displayMessageSessionFinished(String message);
+
+    /**
+     * Return total number of tests in session if known.
+     *
+     * @return number of tests
+     */
+    public abstract int getTotalTests();
+
+    /**
+     * Interface providing factory method for creating {@link TestResultDisplayHandler}s.
+     * Instances should be registered in the default lookup.
+     */
+    public static interface Provider {
```

Comment:
How do you plan to evolve these two types?

I'd suggest to use [singletonizer pattern](http://wiki.apidesign.org/wiki/Singletonizer). E.g.
* make `TestResultDisplayHandler` constructor package private
* add package private `class Impl<T> extends TestResultDisplayHandler` that will have two fields
* `final T data` and `final Provider<T> provider`
* Add parameter `<T>` to `Provider` and let the `create` method `T`
* for each method in `TestResultDisplayHandler` method create the same method in `Provider` with first parameter being `T`.

Here is a simplified example:
```java
public static interface Provider<T> {
  T create(TestSession session);
  int getTotalTests(T data);
}
```
the `Impl` class should then have
```java
class Impl<T> extends TestResultDisplayHandler {
  private final T data;
  private final Provider<T> provider;

  public int getTotalTests() { return provider.getTotalTests(data); }

  static <T> Impl create(Provider<T> p, TestSession s) {
    return new Impl(p, p.create(s));
  }
}
```
the `get` method in the `TestResultDisplayHandler` would then become:
```java
public static final TestResultDisplayHandler get(TestSession session) {
       TestResultDisplayHandler.Provider provider = Lookup.getDefault().lookup(TestResultDisplayHandler.Provider.class);
        if (provider != null) {
            return Impl.create(provider, session);
        }
        return Impl.create(new ResultDisplayHandler());
    }
```

Evolution story would then be:
- add new method into the `TestResultDisplayHandler` (safe as the class isn't publicly subclassable)
- add another `Provider2 extends Provider` SPI interface with the new method(s)

Reply:
Changed as requested.